### PR TITLE
Buffer authority must match upgrade authority for deploys and upgrades

### PR DESCRIPTION
**Pull request by jackcmay**
#### Problem

Buffer account can be recreated after its authority is marked `None`

#### Summary of Changes

- Enforce that buffer authority matches upgrade authority during deploys / upgrades
- Deploys now require an upgrade authority to be specified as a signer
- Buffer state still has an optional authority but restrict everyone from setting it to `None`
- Remove the tooling around buffer write `--final`

The new CLI is also compatible with the existing upgradeable loader and the work flow is the same except:
- Cannot set the buffer's authority to `None` via --final (using either `write-buffer` or `set-buffer-authority`
- When deploying cannot specify the upgrade authority by public key, need to provide a keypair or use the default keypair

Fixes #

### CODE REVIEW

File: sdk/program/src/bpf_loader_upgradeable.rs [hunk 1]
```diff
@@ -103,29 +99,32 @@ pub fn create_buffer(
             UpgradeableLoaderState::buffer_len(program_len)? as u64,
             &id(),
         ),
-        Instruction::new(id(), &UpgradeableLoaderInstruction::InitializeBuffer, metas),
+        Instruction::new(
+            id(),
+            &UpgradeableLoaderInstruction::InitializeBuffer,
+            vec![
+                AccountMeta::new(*buffer_address, false),
+                AccountMeta::new(*authority_address, false),
```
Comment (jstarry) [inline sdk/program/src/bpf_loader_upgradeable.rs:107]:
```suggestion
                AccountMeta::new_readonly(*authority_address, false),
```

File: sdk/program/src/bpf_loader_upgradeable.rs [hunk 2]
```diff
@@ -103,29 +99,32 @@ pub fn create_buffer(
...
-    if let Some(authority_address) = authority_address {
-        metas[1] = AccountMeta::new(*authority_address, true);
-    }
     Instruction::new(
         id(),
         &UpgradeableLoaderInstruction::Write { offset, bytes },
-        metas,
+        vec![
+            AccountMeta::new(*buffer_address, false),
+            AccountMeta::new(*authority_address, true),
```
Comment (jstarry) [inline sdk/program/src/bpf_loader_upgradeable.rs:126]:
```suggestion
            AccountMeta::new_readonly(*authority_address, true),
```

File: runtime/src/loader_utils.rs [hunk 3]
```diff
@@ -89,14 +91,14 @@ pub fn load_buffer_account<T: Client>(
...
                 &buffer_pubkey,
-                None,
+                &buffer_authority_pubkey,
                 offset,
                 chunk.to_vec(),
             )],
             Some(&from_keypair.pubkey()),
         );
         bank_client
-            .send_and_confirm_message(&[from_keypair, &buffer_keypair], message)
```
Comment (jstarry) [inline runtime/src/loader_utils.rs:99]: Oh did the bank client not fail if you try to sign with an unneeded keypair?

File: programs/bpf_loader/src/lib.rs [hunk 4]
```diff
@@ -1927,7 +1981,50 @@ mod tests {
...
-                .send_and_confirm_message(&[&mint_keypair, &program_keypair], message)
+                .send_and_confirm_message(
+                    &[&mint_keypair, &program_keypair, &upgrade_authority_keypair],
+                    message
+                )
+                .unwrap_err()
+                .unwrap()
+        );
+
+        // Mismatched buffer and program authority
```
Comment (jstarry) [inline programs/bpf_loader/src/lib.rs:1992]: Could you also add a test that upgrades/deploys fail when buffer authority is `None`?

**Review by jstarry — DISMISSED** [review]
Looks good!

File: runtime/src/loader_utils.rs (hunk 3 above)
Comment (jackcmay) [reply runtime/src/loader_utils.rs:99]: Writes require the buffer authority, it was just that the buffer authority was the buffer_keypair.
